File: board/board.py
from board.constants import *
from board.move import Move
from board.move_generator import generate_all_moves
from board.rule_checker import filter_legal_moves, is_in_check


class Board:

    def __init__(self):

        self.board = [
            ["bR","bN","bB","bQ","bK","bB","bN","bR"],
            ["bP","bP","bP","bP","bP","bP","bP","bP"],
            ["","","","","","","",""],
            ["","","","","","","",""],
            ["","","","","","","",""],
            ["","","","","","","",""],
            ["wP","wP","wP","wP","wP","wP","wP","wP"],
            ["wR","wN","wB","wQ","wK","wB","wN","wR"]
        ]

        self.turn = WHITE

        self.white_king = (7,4)
        self.black_king = (0,4)

        self.move_history = []

        self.en_passant_target = None

        self.castling_rights = {
            "wK": True,
            "wQ": True,
            "bK": True,
            "bQ": True
        }

    def make_move(self,move):

        move.prev_en_passant = (None if self.en_passant_target is None else tuple(self.en_passant_target))
        move.prev_castling_rights = self.castling_rights.copy()  # ✅ FIX
        move.prev_white_king = self.white_king
        move.prev_black_king = self.black_king

        self.board[move.start_row][move.start_col] = ""
        self.board[move.end_row][move.end_col] = move.piece_moved

        if move.promotion:
            self.board[move.end_row][move.end_col] = move.piece_moved[0] + move.promotion

        # ✅ EN PASSANT
        if move.is_en_passant:
            r, c = move.en_passant_capture_pos
            self.board[r][c] = ""

        # ✅ UPDATE EN PASSANT TARGET
        self.en_passant_target = None
        if move.piece_moved[1] == "P" and abs(move.start_row - move.end_row) == 2:
            mid_row = (move.start_row + move.end_row) // 2
            self.en_passant_target = (mid_row, move.start_col)

        # ✅ CASTLING
        if move.is_castling:
            if move.end_col == 6:  # kingside
                self.board[move.end_row][7] = ""
                self.board[move.end_row][5] = move.piece_moved[0] + "R"
            else:  # queenside
                self.board[move.end_row][0] = ""
                self.board[move.end_row][3] = move.piece_moved[0] + "R"

        # KING UPDATE
        if move.piece_moved == "wK":
            self.white_king = (move.end_row, move.end_col)
        elif move.piece_moved == "bK":
            self.black_king = (move.end_row, move.end_col)

        # CASTLING RIGHTS
        if move.piece_moved == "wK":
            self.castling_rights["wK"] = False
            self.castling_rights["wQ"] = False

        elif move.piece_moved == "bK":
            self.castling_rights["bK"] = False
            self.castling_rights["bQ"] = False

        elif move.piece_moved == "wR":
            if move.start_col == 0:
                self.castling_rights["wQ"] = False
            elif move.start_col == 7:
                self.castling_rights["wK"] = False

        elif move.piece_moved == "bR":
            if move.start_col == 0:
                self.castling_rights["bQ"] = False
            elif move.start_col == 7:
                self.castling_rights["bK"] = False

        self.move_history.append(move)
        self.turn = BLACK if self.turn == WHITE else WHITE


    def undo_move(self):

        if not self.move_history:
            return

        move = self.move_history.pop()

        self.board[move.start_row][move.start_col] = move.piece_moved

        if move.is_en_passant:
            self.board[move.end_row][move.end_col] = ""
            r, c = move.en_passant_capture_pos
            self.board[r][c] = move.piece_captured
        else:
            self.board[move.end_row][move.end_col] = move.piece_captured

        # undo castling rook
        if move.is_castling:
            if move.end_col == 6:
                self.board[move.end_row][5] = ""
                self.board[move.end_row][7] = move.piece_moved[0] + "R"
            else:
                self.board[move.end_row][3] = ""
                self.board[move.end_row][0] = move.piece_moved[0] + "R"

        self.en_passant_target = move.prev_en_passant
        self.castling_rights = move.prev_castling_rights
        self.white_king = move.prev_white_king
        self.black_king = move.prev_black_king

        self.turn = BLACK if self.turn == WHITE else WHITE


    def get_all_legal_moves(self):

        moves = generate_all_moves(self, self.turn)
        return filter_legal_moves(self, moves, self.turn)


    def game_status(self):

        moves = self.get_all_legal_moves()

        if len(moves) == 0:

            if is_in_check(self, self.turn):
                winner = "White" if self.turn == BLACK else "Black"
                return f"{winner} wins by checkmate"
            else:
                return "Draw by stalemate"

        return "ongoing"

File: server-java/src/main/java/com/pace/chess/model/Board.java
package com.pace.chess.model;

import java.util.*;

public class Board {
    public int[] board;
    public int turn; // Piece.WHITE or Piece.BLACK
    public int whiteKing;
    public int blackKing;
    public Map<String, Boolean> castlingRights;
    public Integer enPassantTarget;
    public List<Move> moveHistory;

    public Board() {
        board = new int[128];
        turn = Piece.WHITE;
        whiteKing = 116; // e1 in 0x88 (7*16 + 4)
        blackKing = 4;   // e8 in 0x88 (0*16 + 4)
        castlingRights = new HashMap<>(Map.of("wK",true,"wQ",true,"bK",true,"bQ",true));
        enPassantTarget = null;
        moveHistory = new ArrayList<>();
        // The board array is left empty here because BoardDeserializer will fill it from the API!
    }

    public void makeMove(Move move) {
        move.prevEnPassant      = enPassantTarget;
        move.prevCastlingRights = new HashMap<>(castlingRights);
        move.prevWhiteKing      = whiteKing;
        move.prevBlackKing      = blackKing;

        board[move.startSquare] = Piece.EMPTY;
        board[move.endSquare]   = move.pieceMoved;

        if (move.promotion != 0) {
            int color = move.pieceMoved & (Piece.WHITE | Piece.BLACK);
            board[move.endSquare] = color | move.promotion;
        }

        if (move.isEnPassant && move.enPassantCapturePos != -1) {
            board[move.enPassantCapturePos] = Piece.EMPTY;
        }

        enPassantTarget = null;
        // Pawn double push -> set en passant target
        if ((move.pieceMoved & 7) == Piece.PAWN && Math.abs(move.startSquare - move.endSquare) == 32) {
            enPassantTarget = (move.startSquare + move.endSquare) / 2;
        }

        if (move.isCastling) {
            int color = move.pieceMoved & (Piece.WHITE | Piece.BLACK);
            if (move.endSquare % 16 == 6) { // Kingside
                board[move.endSquare + 1] = Piece.EMPTY;
                board[move.endSquare - 1] = color | Piece.ROOK;
            } else { // Queenside
                board[move.endSquare - 2] = Piece.EMPTY;
                board[move.endSquare + 1] = color | Piece.ROOK;
            }
        }

        if ((move.pieceMoved & 7) == Piece.KING) {
            if ((move.pieceMoved & Piece.WHITE) != 0) whiteKing = move.endSquare;
            else blackKing = move.endSquare;
        }

        // Castling rights invalidation based on square endpoints
        if (move.endSquare == 0) castlingRights.put("bQ", false);
        if (move.endSquare == 7) castlingRights.put("bK", false);
        if (move.endSquare == 112) castlingRights.put("wQ", false);
        if (move.endSquare == 119) castlingRights.put("wK", false);

        if ((move.pieceMoved & 7) == Piece.KING) {
            if ((move.pieceMoved & Piece.WHITE) != 0) { castlingRights.put("wK", false); castlingRights.put("wQ", false); }
            else { castlingRights.put("bK", false); castlingRights.put("bQ", false); }
        } else if ((move.pieceMoved & 7) == Piece.ROOK) {
            if (move.startSquare == 112) castlingRights.put("wQ", false);
            if (move.startSquare == 119) castlingRights.put("wK", false);
            if (move.startSquare == 0) castlingRights.put("bQ", false);
            if (move.startSquare == 7) castlingRights.put("bK", false);
        }

        moveHistory.add(move);
        turn = (turn == Piece.WHITE) ? Piece.BLACK : Piece.WHITE;
    }

    public void undoMove() {
        if (moveHistory.isEmpty()) return;
        Move move = moveHistory.remove(moveHistory.size() - 1);

        board[move.startSquare] = move.pieceMoved;

        if (move.isEnPassant && move.enPassantCapturePos != -1) {
            board[move.endSquare] = Piece.EMPTY;
            board[move.enPassantCapturePos] = move.pieceCaptured;
        } else {
            board[move.endSquare] = move.pieceCaptured;
        }

        if (move.isCastling) {
            int color = move.pieceMoved & (Piece.WHITE | Piece.BLACK);
            if (move.endSquare % 16 == 6) { 
                board[move.endSquare - 1] = Piece.EMPTY; 
                board[move.endSquare + 1] = color | Piece.ROOK; 
            } else { 
                board[move.endSquare + 1] = Piece.EMPTY; 
                board[move.endSquare - 2] = color | Piece.ROOK; 
            }
        }

        enPassantTarget  = move.prevEnPassant;
        castlingRights   = move.prevCastlingRights;
        whiteKing        = move.prevWhiteKing;
        blackKing        = move.prevBlackKing;
        turn = (turn == Piece.WHITE) ? Piece.BLACK : Piece.WHITE;
    }
}